### ISSUE TITLE
NoMethodError: undefined method `next_token_of' for nil:NilClass

### ISSUE THREAD

**Issue by eeisenhart** [rodjek/puppet-lint#906]
puppet-lint version: 2.4.2
ruby version: 2.5.7-p206
platform: x86_64-linux
file path: manifests/client.pp
file contents:
```
class ssu_zabbix::client {

  $zabbixrelease = '3.0'

  iptables::iptable_tcp_port {
    "iptables:tcp:zabbix":
      port   => "10050",
      source => "130.157.30.131",
      ensure => "present";
  }
 
  yumrepo {
    "zabbix-${zabbixrelease}-rhel${facts['os']['release']['major']}-x86_64":
      descr     => 'Zabbix',
      enabled   => 1,
      gpgcheck  => 1,
      baseurl   => "https://repo.zabbix.com/zabbix/${zabbixrelease}/rhel/${facts['os']['release']['major']}/${facts['os']['architecture']}",
      gpgkey    => [ 
        'https://repo.zabbix.com/zabbix-official-repo.key',
        'https://repo.zabbix.com/RPM-GPG-KEY-ZABBIX',
        'https://repo.zabbix.com/RPM-GPG-KEY-ZABBIX-79EA5ED4',
        'https://repo.zabbix.com/RPM-GPG-KEY-ZABBIX-A14FE591',
      ],
  }

  package {
    "zabbix-agent":
      require => Yumrepo["zabbix-${zabbixrelease}-rhel${facts['os']['release']['major']}-x86_64"],
      ensure  => installed;
  }

  file { "/etc/zabbix/zabbix_agentd.conf":
    source  => "puppet:///modules/ssu_zabbix/zabbix_agentd.conf",
    owner   => "root",
    group   => "root",
    mode    => "644",
    ensure  => "present",
  }

  service {
    "zabbix-agent":
     ensure  => running,
     enable  => true,
     require => Package["zabbix-agent"]
  }
}

```
error:
```
NoMethodError: undefined method `next_token_of' for nil:NilClass
/opt/puppetlabs/pdk/share/cache/ruby/2.5.0/gems/puppet-lint-2.4.2/lib/puppet-lint/plugins/check_resources/ensure_first_param.rb:31:in `fix'
/opt/puppetlabs/pdk/share/cache/ruby/2.5.0/gems/puppet-lint-2.4.2/lib/puppet-lint/checkplugin.rb:42:in `block in fix_problems'
/opt/puppetlabs/pdk/share/cache/ruby/2.5.0/gems/puppet-lint-2.4.2/lib/puppet-lint/checkplugin.rb:38:in `each'
/opt/puppetlabs/pdk/share/cache/ruby/2.5.0/gems/puppet-lint-2.4.2/lib/puppet-lint/checkplugin.rb:38:in `fix_problems'
/opt/puppetlabs/pdk/share/cache/ruby/2.5.0/gems/puppet-lint-2.4.2/lib/puppet-lint/checks.rb:67:in `block in run'
/opt/puppetlabs/pdk/share/cache/ruby/2.5.0/gems/puppet-lint-2.4.2/lib/puppet-lint/checks.rb:65:in `each'
/opt/puppetlabs/pdk/share/cache/ruby/2.5.0/gems/puppet-lint-2.4.2/lib/puppet-lint/checks.rb:65:in `run'
/opt/puppetlabs/pdk/share/cache/ruby/2.5.0/gems/puppet-lint-2.4.2/lib/puppet-lint.rb:205:in `run'
/opt/puppetlabs/pdk/share/cache/ruby/2.5.0/gems/puppet-lint-2.4.2/lib/puppet-lint/bin.rb:66:in `block in run'
/opt/puppetlabs/pdk/share/cache/ruby/2.5.0/gems/puppet-lint-2.4.2/lib/puppet-lint/bin.rb:62:in `each'
/opt/puppetlabs/pdk/share/cache/ruby/2.5.0/gems/puppet-lint-2.4.2/lib/puppet-lint/bin.rb:62:in `run'
/opt/puppetlabs/pdk/share/cache/ruby/2.5.0/gems/puppet-lint-2.4.2/bin/puppet-lint:7:in `<top (required)>'
/home/eric/puppet/ssu_zabbix/bin/puppet-lint:29:in `load'
/home/eric/puppet/ssu_zabbix/bin/puppet-lint:29:in `<main>'
```

**Comment by eeisenhart** [comment]
Note: this is the puppet-lint included in pdk-1.14.1

**Comment by usev6** [comment]
Hi @eeisenhart, I'm not able to reproduce the crash with puppet-lint 2.4.2, ruby 2.5.7p206 and the code of ssu_zabbix::client you pasted.

Could you post the command that actual triggers the crash for you? (I've tried with ```puppet-lint --fix modules/ssu_zabbix/manifests/client.pp```.) Also, do you have settings for puppet-lint configured (e.g. in ~/.puppet-lint.rc)?

**Comment by eeisenhart** [comment]
What I actually ran was `pdk validate -a`

No puppet-lint settings of my own.

It worked fine without the “fix” setting and after I manually cleaned
things up, worked fine with the fix option.

-- 
Eric Eisenhart
Lead Linux Sysadmin
He/Him/His

**Comment by usev6** [comment]
Thanks for the feedback. I'm afraid I can't help with this one.

For the record: I install pdk 1.14.1 as a gem, created a module directory with only one file (manifests/client.pp), ran ```pdk convert``` inside the module dir and after that ```pdk validate -a```. ```puppet-lint``` fixed some stuff (like it did when running it on its own), but there was no crash.

**Comment by rodjek** [comment]
OK, I've figured out whats going on here. The underlying problem with `ensure_first_param` is that the fix method is relying on absolute positioning in the `tokens` array rather than operating relative to the tokens in question. This can cause a problem if there are other checks that fix the manifest that run before it that add or remove tokens from the array, throwing the absolute positioning off. Fortunately, this is easily solved.

The reason that it was so hard to reproduce is that the order that the checks are loaded and executed is undefined (which ideally should not matter) and is determined by how the operating systems filesystem globbing works. On my machine (and presumably @usev6's) `ensure_first_param` is loaded and executed before `trailing_whitespace` ,which was the check that was removing tokens from the array for this manifest. On @eeisenhart's machine this is presumably the other way around.

You can forcefully replicate this behaviour with the following script:
```ruby
require 'puppet-lint'

linter = PuppetLint::Checks.new
linter.load_data('test.pp', File.read('test.pp'))

checks = [
  PuppetLint.configuration.check_object[:trailing_whitespace].new,
  PuppetLint.configuration.check_object[:ensure_first_param].new,
]

checks.each { |check| check.run }
checks.each { |check| check.fix_problems }
```

Anyway, long story short, this will be fixed up shortly :)
